src/io/binary_parser.cpp

```diff
@@ -119,6 +109,13 @@ long long read_variables(string filename, dd::FactorGraph &fg) {
     fg.variables[id] =
         dd::Variable(id, type_const, is_evidence, 0, upper_bound, init_value,
                      init_value, edge_count, is_observation);
+
```

Comment:
Note that `lower_bound` seems to be fixed at 0 and `upper_bound` initialized to `cardinality - 1`, then the rest of the code recovers `cardinality` over and over again from these two useless values.